Deprecate DevNull, STDIN, STDOUT, and STDERR to lowercase

ararslan opened the pull request:
EDIT: Initially this PR was just to capitalize `DevNull` but the intent has since changed. The text below summarizes the initial state of the PR.

---

This is consistent with the other stream constants, i.e. `STDOUT`, `STDIN`, and `STDERR`, and indeed this was a violation of the typical style of all-caps constants.

Fixes #25786.

vtjnash commented:
> was a violation of the typical style of all-caps constants.

In that case, shouldn't we make all of these lower-caps instead? (that's how they are in C, for example: stdin / stdout / stderr / stdio / devnull).

ararslan commented:
> In that case, shouldn't we make all of these lower-caps instead?

I don't understand what you mean. We don't do that for anything else, and these aren't just for C compatibility.

JeffBezanson commented on base/exports.jl:
```diff
@@ -40,7 +40,7 @@ export
     DenseVecOrMat,
     DenseVector,
-    DevNull,
+    DEVNULL,
```
Should move this to the constants block alongside `STDOUT` etc.

StefanKarpinski commented:
Most of our global singletons like `nothing` and `missing` are lowercase. There's also the fact that this looks quite suggestive of what it does:
```jl
pipeline(`echo Hello`, stdout=stderr)
```
i.e. temporarily replace `stdout` with the value of `stderr`. I'm in favor of changing these all to lowercase: `stdin`, `stdout`, `stderr`, `devnull`.

ararslan commented:
In the particular example of `pipeline`, you run into the issue where the keyword argument name shadows a global:
```
julia> const x = 1
1

julia> f(x=4, y=x) = (x, y)
f (generic function with 3 methods)

julia> f()
(4, 4)
```
Note that the keyword argument value of `x` is used rather than the global `x`. In `pipeline`, it's possible to specify both `stdout` and `stderr`, so without constants that can be differentiated by name from those, everything will get all messed up.

JeffBezanson commented:
That's only an issue for the method definition, not call sites. In the method definition these default to `nothing`, so there should be no problem.

JeffBezanson reviewed (changes requested):
I think there's still some controversy here over whether we want lowercase or uppercase.

ararslan committed:
Adjust some uses of std(in|out|err) as a variable name

This includes bumping the Documenter requirement to 0.13.2, which
contains similar changes for usages of these names as variables

ararslan committed:
Deprecate DevNull to devnull

ararslan committed:
Rename STD(IN|OUT|ERR) to their lowercase equivalents

ararslan commented:
Per triage consensus, this now deprecates `DevNull`, `STDIN`, `STDOUT`, and `STDERR` to `devnull`, `stdin`, `stdout`, and `stderr`, respectively. I did this in three commits: one to change uses of the lowercase versions as local variables, one to deprecate `DevNull`, and one to deprecate the others.

StefanKarpinski commented:
I find this to be rather pleasant. Thanks for doing it despite dissenting, Alex! In 1.0 we could rename DevNullStream to DevNull. Not sure if that’s even a public name.

ararslan commented:
Hm, looks like there's an issue building the docs. I'm not sure what the deal is yet.

ararslan commented:
Okay, doc building is apparently still an issue, even after updating Documenter.

mortenpi commented:
In case you already didn't come to the same conclusion: [`withoutput`](https://github.com/JuliaDocs/Documenter.jl/blob/master/src/Utilities/Utilities.jl#L558-L604) crashes when evaluating the following REPL block:

````
```@repl
versioninfo()
```
````

It works fine on `master` though, so it's something in this PR.

ararslan commented:
Hm, somewhere it's using `Core.stdout` where it should be using `Base.stdout`, which is a `LibuvStream` and thus can be passed to `redirect_stdout`.

JeffBezanson commented:
This is because binding deprecations only work with constants. `Base.stdout` changes at some point, and `Base.STDOUT` continues to point to the value it had when deprecated.jl ran.

The simplest fix is to also update the old names in reinit_stdio in libuv.jl. But of course that won't handle any other cases of changing the variables.

vtjnash commented:
> But of course that won't handle any other cases of changing the variables.

The only other place they can be changed is via the `redirect_std*` functions defined in stream.jl

ararslan commented:
Can someone explain why `Core.stdout` and `Base.stdout` are and need to be distinct?

JeffBezanson commented:
It's so some basic printing functionality can be available before Base is defined.

ararslan commented:
Okay, makes sense, thanks. I figured it would have to be something like that.

Could we make `Base.stdout` et al. `const`s? Is there a reason why they aren't currently?

JeffBezanson commented:
They are changed to acquire more functionality after bootstrap, and they're changed by the `redirect_` functions.

JeffBezanson commented:
Are you close here or should I make the remaining change?

JeffBezanson commented:
Will do.

fredrikekre commented:
```julia
julia> print(STDOUT, "foo")
WARNING: Base.STDOUT is deprecatednothing
 in module Main
foo
```

stevengj commented:
(Unfortunately, this PR is going to cause an huge number of deprecation warnings, and just switching to lowercase and doing `using Compat` will not be sufficient.  People will need to put `@compat` decorators everywhere they use the uppercase names.)

StefanKarpinski commented:
Why can't `Compat` export `stdin`, `stdout`, `stderr` and `devnull` as aliases? None of those names are exported or defined by `Base` in any earlier version of Julia.

KristofferC commented:
Because functions like `redirect_stdout` rebinds the variables.

StefanKarpinski commented:
That is annoying. I'm not sure what we can do about it though.

KristofferC commented:
I thought about using `getproperty`-overloading but that is not available on 0.6 :(

StefanKarpinski commented:
How about making `stdout` et al. mutable wrappers around an io object that just proxies IO calls through. Then the binding can be `const` and to redirect it, you just put a different IO object inside the wrapper. That definition could be put in Base and copied in Compat.

mbauman commented:
Seems like we could have Compat override the `Base._redirect_*` methods that actually do the rebindings.  They're totally untyped so we might even be able to do it without a warning with a `::IO` annotation.

ararslan commented:
This has the critically unfortunate consequence of conflicting with `StatsBase.stderr`.

stevengj commented:
Note that the SciPy [`stats.stderr` function was deprecated in favor of `stats.sem`](https://docs.scipy.org/doc/scipy-0.8.x/reference/generated/scipy.stats.mstats.stderr.html).

sbromberger commented:
Sorry for jumping in here so late, but what about `Random.GLOBAL_RNG`?

KristofferC commented:
That is a constant so how does the discussion here apply?

sbromberger commented:
ah, I see:

```
const GLOBAL_RNG = MersenneTwister(0)
```

Thanks.
